### ISSUE TITLE
Ansible 2.1 overrides default variables by accident

### ISSUE THREAD

**Issue by ThomasSteinbach** [ansible/ansible#14296]
Since I recently updated my Ansible installation repository, Ansible overrides default variables where it should not. If I call a sub role as dependency of a 'parent' role twice, but the last time with overriding a default variable, the variable was also overridden on the first call. Here is a tiny example setup (I omitted the starting `---` in every file):

**sub_role/defaults/main.yml**

```
myvar: "first call"
```

**sub_role/tasks/main.yml**

```
- debug: var=myvar
```

**parent_role/meta/main.yml**

```
dependencies:
  - { role: sub_role }
  - { role: sub_role, myvar: "second call" }
```

Now executing the _parent_role_ leads to following wrong result:

```
TASK [sub_role : debug] ********************************************************
ok: [localhost] => {
    "myval": "second call"
}

TASK [sub_role : debug] ********************************************************
ok: [localhost] => {
    "myval": "second call"
}

```

Can you see the error? On the first call, the output of `myval` should be _first call_ and not _second call_. Because of recent changes Ansible seems to interpret all role variables, starting with the first definition and ignoring omitted definitions.

**Comment by dagwieers** [comment]
I can confirm this behaviour. If you run the `sub_role` twice from a playbook, this behaviour does not occur. But if you run it from another rol (as reported) the exact same thing happens.

``` yaml
- hosts: localhost
  roles:
  - role: sub_role
  - role: sub_role                                                                                                                                          
    myvar: second call
```

Results in:

```
[dag@moria ansible.dag]$ ansible-playbook test50.yml 

PLAY ***************************************************************************

TASK [sub_role : debug] ********************************************************
ok: [localhost] => {
    "myvar": "first call"
}

TASK [sub_role : debug] ********************************************************
ok: [localhost] => {
    "myvar": "second call"
}

PLAY RECAP *********************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=0   
```

But if you run the `sub_role` twice from a `parent_role`:

``` yaml
- hosts: localhost
  gather_facts: no                                                                                                                                          
  roles:
  - role: parent_role
```

It fails like this:

```
[dag@moria ansible.dag]$ ansible-playbook test51.yml 

PLAY ***************************************************************************

TASK [sub_role : debug] ********************************************************
ok: [localhost] => {
    "myvar": "second call"
}

TASK [sub_role : debug] ********************************************************
ok: [localhost] => {
    "myvar": "second call"
}

PLAY RECAP *********************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=0   
```

**Comment by jjshoe** [comment]
I see this behavior in 1.9.4 as well. If I use vars/ instead of defaults/ all is well. Though my test case involves using meta to have different roles bring in the same third role.

**Comment by jjshoe** [comment]
Related to #14017?

**Comment by jjshoe** [comment]
Here's an example playbook to use: https://github.com/jjshoe/ansible-14296-test

A and B print correctly as A and B
C and D print incorrectly as D and D
E and F print correctly as E and F
G and H don't print at all

I would expect the behavior of C and D to work like A and B. E and F work as you define the variable being passed into the dependent role. I suspect G and H don't work as the roles defaults/main.yml hasn't been processed yet, and I assume that's correct as well, was hoping to use it as a workaround for C and D not printing correctly. 

Confirmed behavior is same between 1.9.4 and 2.0.0.2.

**Comment by fabianfreyer** [comment]
Created another [minimal working example](https://github.com/fabianfreyer/ansible-dependency-variable-scoping-test).

For the first play in the playbook, this is the expected behaviour:

```
PLAY ***************************************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [dependency : debug] ******************************************************
ok: [localhost] => {
    "foo": "baz"
}

TASK [dependency : debug] ******************************************************
ok: [localhost] => {
    "foo": "bar"
}

PLAY RECAP *********************************************************************
localhost                  : ok=3    changed=0    unreachable=0    failed=0   
```

Actual behaviour

```
PLAY ***************************************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [dependency : debug] ******************************************************
ok: [localhost] => {
    "foo": "baz"
}

TASK [dependency : debug] ******************************************************
ok: [localhost] => {
    "foo": "baz"
}

PLAY RECAP *********************************************************************
localhost                  : ok=3    changed=0    unreachable=0    failed=0   
```

The second play includes the dependency role directly and works as expected.

**Comment by sloonz** [comment]
Duplicate of  #14017 ?

**Comment by jimi-c** [comment]
@fabianfreyer after fixing #14472/#14475 this morning, your example works as expected:

<pre>
# ansible-playbook -vv playbook.yml -i inventory 
Using /etc/ansible/ansible.cfg as config file
PLAYBOOK: playbook.yml *********************************************************
2 plays in playbook.yml
PLAY [all] *********************************************************************
TASK [setup] *******************************************************************
ok: [localhost]
TASK [dependency : debug] ******************************************************
task path: /root/testing/14296/ansible-dependency-variable-scoping-test/roles/dependency/tasks/main.yml:3
ok: [localhost] => {
    "foo": "baz"
}
TASK [dependency : debug] ******************************************************
task path: /root/testing/14296/ansible-dependency-variable-scoping-test/roles/dependency/tasks/main.yml:3
ok: [localhost] => {
    "foo": "bar"
}
PLAY [all] *********************************************************************
TASK [setup] *******************************************************************
ok: [localhost]
TASK [dependency : debug] ******************************************************
task path: /root/testing/14296/ansible-dependency-variable-scoping-test/roles/dependency/tasks/main.yml:3
ok: [localhost] => {
    "foo": "baz"
}
TASK [dependency : debug] ******************************************************
task path: /root/testing/14296/ansible-dependency-variable-scoping-test/roles/dependency/tasks/main.yml:3
ok: [localhost] => {
    "foo": "bar"
}
PLAY RECAP *********************************************************************
localhost                  : ok=6    changed=0    unreachable=0    failed=0   
</pre>

It also appears that https://github.com/ansible/ansible/issues/14017 has been resolved by this work.

However, @jjshoe's example fails still as he showed (plus now I see a recursive templating error at the end with the `- debug: var=name`). Still looking into that.

**Comment by jimi-c** [comment]
So I have the C/D example from @jjshoe's example working, however I'm not sure if it's a "correct" fix. The reason this is happening because only the current role's defaults override all other roles, but role `z` has no defaults, so `d` (being the last role read) wins. The solution is to have the role defaults look up the dependency chain, which means `z` would also grab `c`'s defaults when compiling its view of the defaults.

The reason I'm not sure if this is correct is because we've never done it this way - as @jjshoe points out this result is the same as it has been since 1.9.4 (and most likely earlier). So there is risk here of breaking things because of a precedence change.

Here's the patch, if anyone would like to test it:

<pre>
diff --git a/lib/ansible/playbook/role/__init__.py b/lib/ansible/playbook/role/__init__.py
index 133dd50..b86ad0f 100644
--- a/lib/ansible/playbook/role/__init__.py
+++ b/lib/ansible/playbook/role/__init__.py
@@ -252,10 +252,13 @@ class Role(Base, Become, Conditional, Taggable):
     def get_parents(self):
         return self._parents
 
-    def get_default_vars(self):
+    def get_default_vars(self, dep_chain=[]):
         default_vars = dict()
         for dep in self.get_all_dependencies():
             default_vars = combine_vars(default_vars, dep.get_default_vars())
+        if dep_chain:
+            for parent in dep_chain:
+                default_vars = combine_vars(default_vars, parent._default_vars)
         default_vars = combine_vars(default_vars, self._default_vars)
         return default_vars
 
diff --git a/lib/ansible/vars/__init__.py b/lib/ansible/vars/__init__.py
index 2d1a872..c95d25c 100644
--- a/lib/ansible/vars/__init__.py
+++ b/lib/ansible/vars/__init__.py
@@ -220,7 +220,7 @@ class VariableManager:
             # sure it sees its defaults above any other roles, as we previously
             # (v1) made sure each task had a copy of its roles default vars
             if task and task._role is not None:
-                all_vars = combine_vars(all_vars, task._role.get_default_vars())
+                all_vars = combine_vars(all_vars, task._role.get_default_vars(dep_chain=task._block._dep_chain))
 
         if host:
             # next, if a host is specified, we load any vars from group_vars
</pre>

**Comment by jimi-c** [comment]
I've gone ahead and merged in the above commit, so all problems reported on this ticket should now be addressed in the devel branch. The last fix will be in the next release, but the others should be included in 2.0.2.

If you continue seeing any problems related to this issue, or if you have any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular issue is resolved.

Thank you!

**Comment by sprhawk** [comment]
Is this still same thing? I'm using ansible 2.7.0, but it seems still to be here.

To my understand to [official doc](https://docs.ansible.com/ansible/2.3/playbooks_variables.html#variable-precedence-where-should-i-put-a-variable)

any variable defined for a role will override defaults "GLOBALLY" ?

so for other I just cannot use the defaults anymore?
